tests/codegen/read-noundef-metadata.rs

```diff
@@ -0,0 +1,51 @@
+// compile-flags: -O -Z merge-functions=disabled
+// no-system-llvm
+// ignore-debug (the extra assertions get in the way)
+
+#![crate_type = "lib"]
+
+// Ensure that various forms of reading pointers correctly annotate the `load`s
+// with `!noundef` metadata to enable extra optimization.  The functions return
+// `MaybeUninit` to keep it from being inferred from the function type.
```

Comment:
But doesn't `MaybeUninit::new(...)` call imply `!noundef` of its argument anyway?

Reply:
Good point.  I'll remove the wrapping; the tests pass without it -- I was just trying (and failing) to be safe for the future.

Reply:
Also updated the codegen test to add a second set using `char` to test `!range`+`!nounwind` as well.